storage_controller/src/service.rs

```diff
@@ -6179,4 +6268,95 @@ impl Service {
 
         global_observed
     }
+
+    /// Collect the details for the current proccess wishing to become the storage controller
+    /// leader.
+    ///
+    /// On failures to discover and resolve the hostname the process is killed and we rely on k8s to retry.
+    fn get_proposed_leader_info(&self) -> LeaderPersistence {
+        let hostname = match dns_lookup::get_hostname() {
```

Comment:
It's a bit fragile to use a node's self-perceived DNS name when advertising an address to other nodes: most systems end up needing some kind of configuration to tell a node it's "public name" in case it's running in an environment where the local DNS lookup doesn't give a nice answer.

However, if this works in k8s it's probably good enough for now: might want to add a comment here about the assumption though., and perhaps an assertion that the hostname we get here isn't something silly like `localhost` (or if we _do_ hit `localhost` in tests, then at least warn on it in case someone tries this on a self hosted system)

Reply:
Done in https://github.com/neondatabase/neon/pull/8588/commits/2a16848c94125307b63bfaf5df95b11aad9f056d. We now take a cli arg `address-for-peers` (open to better names) instead of inferring it from via dns lookup.